### PR TITLE
Issue2605 - MMS download needs to stream object data into a file inst…

### PR DESCRIPTION
**Pull request by LiilyZhang**
…ead of save in memory before writing to a file

Signed-off-by: zhangl <zhangl@us.ibm.com>

### CODE REVIEW

File: cli/sync_service/download.go [hunk 1]
```diff
@@ -109,23 +90,49 @@ func ObjectDownLoad(org string, userPw string, objType string, objId string, fil
...
+			}
+			msgPrinter.Println("Verifying digital signature is done.")
+			rename = true
+		}
 
-	if _, err := file.Write(data); err != nil {
-		cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to save data for object '%s' of type '%s' to file %s", objId, objType, fileName))
+	}
+
+	if rename {
```
Comment (dabooz) [inline cli/sync_service/download.go:117]: Can you move this rename logic into the VerifyDataSig function?

File: cli/sync_service/download.go [hunk 2]
```diff
@@ -109,23 +90,49 @@ func ObjectDownLoad(org string, userPw string, objType string, objId string, fil
...
+		// rename fileName.tmp
+                tmpFileName := fmt.Sprintf("%s.tmp", fileName)
+                if err := os.Rename(tmpFileName, fileName); err != nil {
+                        cliutils.Fatal(cliutils.INTERNAL_ERROR, msgPrinter.Sprintf("Failed to rename file '%s' to '%s', err: %v", tmpFileName, fileName, err))
+                }
+	} else {
```
Comment (dabooz) [inline cli/sync_service/download.go:123]: This else clause would become the else of the if !skipDigitalVerify

**Review by dabooz — CHANGES_REQUESTED** [review]

File: cli/sync_service/download.go (hunk 2 above)
Comment (LiilyZhang) [reply cli/sync_service/download.go:123]: Yes, that is right. So 
if !skipDigitalVerify, 
     the data will be streamed to a tmp file, at same time generate hash + verify.
then rename the tmp file
else 
    the data is stream to the file directly

Comment (LiilyZhang) [reply cli/sync_service/download.go:123]: @dabooz ah I think I know what you mean. Yea after moving the rename logic, will also change this part

File: cli/sync_service/download.go [hunk 3]
```diff
@@ -157,7 +168,56 @@ func VerifyDataSig(dataReader io.Reader, publicKey string, signature string, has
...
+}
+
+func (wc *WriteCounter) Write(p []byte) (n int, err error) {
+	n, err = wc.PassInWriter.Write(p)
+	wc.ChunkNumber += 1
+	wc.PrintProgress(n)
+	return n, err
+}
+
+func (wc *WriteCounter) PrintProgress(n int) {
```
Comment (dabooz) [inline cli/sync_service/download.go:216]: I really dont like how the dots work. It would look better if we could use the % counter like we do for upload.

**Review by dabooz — CHANGES_REQUESTED** [review]

**Review by dabooz — APPROVED** [review]

**Pull request merged**